From d9ec151d65fc0bd99fc9265f183cf3d057cdec68 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Thu, 8 Oct 2026 06:19:11 +0200
Subject: [PATCH] Location of the arrays of a material data manager

The location of the arrays (MemoryLocation: host, device or shared) is a
property of the data, declared by the initializer of a material data
manager and checked:

- a behaviour generated by the generic-parallel interface with the cuda or
  hip backend refuses data located in the memory of the host;
- when the arrays are not located in the memory of the host, all the arrays
  of the states must be given by the initializer, and the arrays of the
  tangent operators and of the speed of sound must be given if they are
  used: the arrays allocated by the manager are in the memory of the host;
- update and revert, which copy the states on the host, refuse data located
  in the memory of a device.

The headers BatchExecutionPolicy.hxx and MemoryLocation.hxx are installed
(the first one was missing).

GenericParallelCUDATest declares the managed memory as shared and checks
each refusal by its message.
---
 include/CMakeLists.txt                        |  2 +
 .../MGIS/Behaviour/MaterialDataManager.hxx    | 11 ++++
 include/MGIS/Behaviour/MemoryLocation.hxx     | 39 ++++++++++++
 src/Integrate.cxx                             | 20 ++++++
 src/MaterialDataManager.cxx                   | 62 ++++++++++++++++++
 tests/GenericParallelCUDATest.cxx             | 63 ++++++++++++++++---
 6 files changed, 189 insertions(+), 8 deletions(-)
 create mode 100644 include/MGIS/Behaviour/MemoryLocation.hxx

diff --git a/include/CMakeLists.txt b/include/CMakeLists.txt
index 60881d9..f2f4fb8 100644
--- a/include/CMakeLists.txt
+++ b/include/CMakeLists.txt
@@ -42,6 +42,8 @@ mgis_header(MGIS/Behaviour FiniteStrainBehaviourOptions.hxx)
 mgis_header(MGIS/Behaviour BehaviourFctPtr.hxx)
 mgis_header(MGIS/Behaviour BatchDataView.hxx)
 mgis_header(MGIS/Behaviour BatchBehaviourFctPtr.hxx)
+mgis_header(MGIS/Behaviour BatchExecutionPolicy.hxx)
+mgis_header(MGIS/Behaviour MemoryLocation.hxx)
 mgis_header(MGIS/Behaviour BehaviourDescription.hxx)
 mgis_header(MGIS/Behaviour Behaviour.hxx)
 mgis_header(MGIS/Behaviour StateView.hxx)
diff --git a/include/MGIS/Behaviour/MaterialDataManager.hxx b/include/MGIS/Behaviour/MaterialDataManager.hxx
index 2abf4dd..685d4ff 100644
--- a/include/MGIS/Behaviour/MaterialDataManager.hxx
+++ b/include/MGIS/Behaviour/MaterialDataManager.hxx
@@ -26,6 +26,7 @@
 #include "MGIS/Config.hxx"
 #include "MGIS/Behaviour/MaterialStateManager.hxx"
 #include "MGIS/Behaviour/BatchExecutionPolicy.hxx"
+#include "MGIS/Behaviour/MemoryLocation.hxx"
 
 namespace mgis::behaviour {
 
@@ -97,6 +98,14 @@ namespace mgis::behaviour {
      * end of the time step.
      */
     MaterialStateManagerInitializer s1;
+    /*!
+     * \brief location of the arrays given by this initializer. If it is not
+     * the memory of the host, all the arrays of the states and the arrays of
+     * the tangent operators and of the speed of sound, if they are used, must
+     * be given: the arrays allocated by the material data manager are in the
+     * memory of the host.
+     */
+    MemoryLocation location = MemoryLocation::HOST;
   };  // end of MaterialDataManagerInitializer
 
   /*!
@@ -242,6 +251,8 @@ namespace mgis::behaviour {
     std::span<real> speed_of_sound;
     //! \brief number of integration points
     const size_type n;
+    //! \brief location of the arrays (see `MaterialDataManagerInitializer`)
+    const MemoryLocation memory_location = MemoryLocation::HOST;
     /*!
      * \brief the size of the stiffness matrix for one integration point (the
      * size of K is K_stride times the number of integration points)
diff --git a/include/MGIS/Behaviour/MemoryLocation.hxx b/include/MGIS/Behaviour/MemoryLocation.hxx
new file mode 100644
index 0000000..27401d5
--- /dev/null
+++ b/include/MGIS/Behaviour/MemoryLocation.hxx
@@ -0,0 +1,39 @@
+/*!
+ * \file   include/MGIS/Behaviour/MemoryLocation.hxx
+ * \brief  This file declares the `MemoryLocation` enumeration.
+ * \date   09/10/2026
+ * \copyright Use, modification and distribution are subject
+ * to one of the following licences:
+ * - GNU Lesser General Public License (LGPL), Version 3.0. (See accompanying
+ *   file LGPL-3.0.txt)
+ * - CECILL-C,  Version 1.0 (See accompanying files
+ *   CeCILL-C_V1-en.txt and CeCILL-C_V1-fr.txt).
+ */
+
+#ifndef LIB_MGIS_BEHAVIOUR_MEMORYLOCATION_HXX
+#define LIB_MGIS_BEHAVIOUR_MEMORYLOCATION_HXX
+
+namespace mgis::behaviour {
+
+  /*!
+   * \brief location of the arrays of a material data manager, declared by
+   * the code which allocates them. It is a property of the data, checked
+   * against the backend of the behaviours generated by the
+   * `generic-parallel` interface of `MFront` and against the functions which
+   * access the arrays on the host.
+   */
+  enum struct MemoryLocation {
+    //! \brief ordinary memory of the host
+    HOST,
+    //! \brief memory of a device, which is not accessible by the host
+    DEVICE,
+    /*!
+     * \brief memory accessible by the host and by the device (managed
+     * memory, shared memory of SYCL, etc.)
+     */
+    SHARED
+  };
+
+}  // end of namespace mgis::behaviour
+
+#endif /* LIB_MGIS_BEHAVIOUR_MEMORYLOCATION_HXX */
diff --git a/src/Integrate.cxx b/src/Integrate.cxx
index ce6fd7d..19dc505 100644
--- a/src/Integrate.cxx
+++ b/src/Integrate.cxx
@@ -445,6 +445,23 @@ namespace mgis::behaviour::internals {
     return Field{values.data() + b * stride, stride, 1};
   }  // end of makeBatchField
 
+  /*!
+   * \brief check that the arrays of a material data manager are accessible
+   * by the device of the backend of its behaviour
+   * \param[in] m: material data manager
+   */
+  static void checkBatchMemoryLocation(const MaterialDataManager& m) {
+    if ((m.memory_location == MemoryLocation::HOST) &&
+        (isDeviceOnlyBackend(m.b))) {
+      mgis::raise(
+          "the arrays of the material data manager are located in the memory "
+          "of the host, which is not accessible by the device of the backend "
+          "'" +
+          m.b.backend +
+          "' (see the location member of MaterialDataManagerInitializer)");
+    }
+  }  // end of checkBatchMemoryLocation
+
   /*!
    * \brief view on the points `[b, e)` of a material data manager, as a
    * batch (`generic-parallel` interface), and fields of the material
@@ -559,6 +576,7 @@ namespace mgis::behaviour::internals {
           "integrate: the behaviour has not been generated by the "
           "generic-parallel interface");
     }
+    checkBatchMemoryLocation(m);
     if (b == e) {
       return BehaviourIntegrationResult{};
     }
@@ -596,6 +614,7 @@ namespace mgis::behaviour::internals {
       const ConstField* const inputs,
       const size_type b,
       const size_type e) {
+    checkBatchMemoryLocation(m);
     if (b == e) {
       return BehaviourIntegrationResult{};
     }
@@ -618,6 +637,7 @@ namespace mgis::behaviour::internals {
       const BehaviourPostProcessing& f,
       const size_type b,
       const size_type e) {
+    checkBatchMemoryLocation(m);
     if (b == e) {
       return BehaviourIntegrationResult{};
     }
diff --git a/src/MaterialDataManager.cxx b/src/MaterialDataManager.cxx
index f5f58fd..a43f806 100644
--- a/src/MaterialDataManager.cxx
+++ b/src/MaterialDataManager.cxx
@@ -57,8 +57,33 @@ namespace mgis::behaviour {
       : s0(behaviour, s, i.s0),
         s1(behaviour, s, i.s1),
         n(s),
+        memory_location(i.location),
         K_stride(getTangentOperatorArraySize(behaviour)),
         b(behaviour) {
+    if (this->memory_location != MemoryLocation::HOST) {
+      // the arrays allocated by the state managers are in the memory of the
+      // host
+      const auto h = behaviour.hypothesis;
+      for (const auto& si : {i.s0, i.s1}) {
+        const auto missing =
+            ((si.gradients.empty()) &&
+             (getArraySize(behaviour.gradients, h) != 0)) ||
+            ((si.thermodynamic_forces.empty()) &&
+             (getArraySize(behaviour.thermodynamic_forces, h) != 0)) ||
+            ((si.internal_state_variables.empty()) &&
+             (getArraySize(behaviour.isvs, h) != 0)) ||
+            ((si.stored_energies.empty()) &&
+             (behaviour.computesStoredEnergy)) ||
+            ((si.dissipated_energies.empty()) &&
+             (behaviour.computesDissipatedEnergy));
+        if (missing) {
+          mgis::raise(
+              "MaterialDataManager::MaterialDataManager: all the arrays of "
+              "the states must be given by the initializer when they are not "
+              "located in the memory of the host");
+        }
+      }
+    }
     if (!i.K.empty()) {
       this->useExternalArrayOfTangentOperatorBlocks(i.K);
     }
@@ -90,7 +115,25 @@ namespace mgis::behaviour {
     this->thread_safe = bv;
   }  // end of setThreadSafe
 
+  /*!
+   * \brief check that an array can be allocated by a material data manager,
+   * i.e. in the memory of the host
+   * \param[in] m: material data manager
+   * \param[in] v: array
+   * \param[in] n: name of the array
+   */
+  static void checkHostAllocation(const MaterialDataManager& m,
+                                  const std::span<real>& v,
+                                  const char* const n) {
+    if ((v.empty()) && (m.memory_location != MemoryLocation::HOST)) {
+      mgis::raise("MaterialDataManager: the array of the " + std::string{n} +
+                  " must be given by the initializer, since the arrays are "
+                  "not located in the memory of the host");
+    }
+  }  // end of checkHostAllocation
+
   void MaterialDataManager::allocateArrayOfTangentOperatorBlocks() {
+    checkHostAllocation(*this, this->K, "tangent operators");
     if (this->thread_safe) {
       allocateArrayWithSynchronization(this->K, this->K_values,
                                        this->n * this->K_stride);
@@ -117,6 +160,7 @@ namespace mgis::behaviour {
   }  // end of useExternalArrayOfTangentOperatorBlocks
 
   void MaterialDataManager::allocateArrayOfSpeedOfSounds() {
+    checkHostAllocation(*this, this->speed_of_sound, "speed of sound");
     if (this->thread_safe) {
       allocateArrayWithSynchronization(this->speed_of_sound,
                                        this->speed_of_sound_values, this->n);
@@ -167,7 +211,24 @@ namespace mgis::behaviour {
 
   MaterialDataManager::~MaterialDataManager() = default;
 
+  /*!
+   * \brief check that the arrays of a material data manager are accessible
+   * by the host
+   * \param[in] f: calling function
+   * \param[in] m: material data manager
+   */
+  static void checkHostAccess(const char* const f,
+                              const MaterialDataManager& m) {
+    if (m.memory_location == MemoryLocation::DEVICE) {
+      mgis::raise(std::string{f} +
+                  ": the arrays of the material data manager are located in "
+                  "the memory of a device, which is not accessible by the "
+                  "host");
+    }
+  }  // end of checkHostAccess
+
   void update(MaterialDataManager& m, const UpdateArguments& args) {
+    checkHostAccess("update", m);
     if (args.zero_tangent_operator_blocks) {
       std::fill(m.K.begin(), m.K.end(), real{0});
     }
@@ -179,6 +240,7 @@ namespace mgis::behaviour {
   }  // end of update
 
   void revert(MaterialDataManager& m) {
+    checkHostAccess("revert", m);
     std::fill(m.K.begin(), m.K.end(), real{0});
     updateValues(m.s1, m.s0);
   }  // end of update
diff --git a/tests/GenericParallelCUDATest.cxx b/tests/GenericParallelCUDATest.cxx
index f1af800..102d43e 100644
--- a/tests/GenericParallelCUDATest.cxx
+++ b/tests/GenericParallelCUDATest.cxx
@@ -42,6 +42,20 @@ static bool check(const bool b, const std::string_view msg) {
   return b;
 }
 
+/*!
+ * \return true if the given function throws an exception whose message
+ * contains the given text
+ */
+template <typename Function>
+static bool throws(const Function& f, const std::string_view text) {
+  try {
+    f();
+  } catch (std::exception& e) {
+    return std::string_view{e.what()}.find(text) != std::string_view::npos;
+  }
+  return false;
+}
+
 //! \brief array of values in managed memory
 struct ManagedArray {
   explicit ManagedArray(const mgis::size_type s) : size(s) {
@@ -98,6 +112,7 @@ struct ManagedData {
     i.K = this->K.span();
     i.s0 = this->s0.initializer();
     i.s1 = this->s1.initializer();
+    i.location = mgis::behaviour::MemoryLocation::SHARED;
     return i;
   }
   ManagedArray K;
@@ -227,17 +242,48 @@ static void testElasticity(const std::string& library,
   setMaterialProperty(managers.gpu_m.s1, "PoissonRatio", 0.3);
   auto opts = BehaviourIntegrationOptions{};
   check(
-      [&managers, &opts] {
-        try {
-          static_cast<void>(integrate(managers.gpu_m, opts, 1));
-        } catch (...) {
-          return true;
-        }
-        return false;
-      }(),
+      throws([&managers,
+              &opts] { static_cast<void>(integrate(managers.gpu_m, opts, 1)); },
+             "must be given as a span"),
       "a uniform value in the memory of the host must be refused");
 }  // end of testElasticity
 
+//! \brief the location of the data is checked against the backend
+static void testMemoryLocations(const std::string& gpu_library) {
+  using namespace mgis::behaviour;
+  constexpr auto n = mgis::size_type{10};
+  const auto b = load(gpu_library, "Elasticity", Hypothesis::TRIDIMENSIONAL);
+  const auto opts = BehaviourIntegrationOptions{};
+  // data in the memory of the host
+  auto m = MaterialDataManager(b, n);
+  check(throws([&m, &opts] { static_cast<void>(integrate(m, opts, 1)); },
+               "located in the memory of the host"),
+        "data in the memory of the host must be refused");
+  // missing arrays
+  auto data = ManagedData(b, n);
+  auto i = data.initializer();
+  i.s1.gradients = {};
+  check(throws([&b, &i] { MaterialDataManager(b, n, i); },
+               "must be given by the initializer"),
+        "missing arrays must be refused");
+  // missing tangent operators
+  auto i2 = data.initializer();
+  i2.K = {};
+  auto m2 = MaterialDataManager(b, n, i2);
+  auto opts2 = BehaviourIntegrationOptions{};
+  opts2.integration_type =
+      IntegrationType::INTEGRATION_CONSISTENT_TANGENT_OPERATOR;
+  check(throws([&m2, &opts2] { static_cast<void>(integrate(m2, opts2, 1)); },
+               "tangent operators must be given"),
+        "missing tangent operators must be refused");
+  // update on the host of data located on the device
+  auto i3 = data.initializer();
+  i3.location = MemoryLocation::DEVICE;
+  auto m3 = MaterialDataManager(b, n, i3);
+  check(throws([&m3] { update(m3); }, "not accessible by the host"),
+        "the update on the host of data located on a device must be refused");
+}  // end of testMemoryLocations
+
 int main(const int argc, const char* const* const argv) {
   if (argc != 3) {
     std::cerr << "usage: " << argv[0] << " cuda-library generic-library\n";
@@ -246,6 +292,7 @@ int main(const int argc, const char* const* const argv) {
   try {
     testNorton(argv[2], argv[1]);
     testElasticity(argv[2], argv[1]);
+    testMemoryLocations(argv[1]);
   } catch (std::exception& e) {
     std::cerr << e.what() << '\n';
     return EXIT_FAILURE;
-- 
2.43.0

